components/style/stylist.rs

```diff
@@ -1524,86 +1465,108 @@ impl SelectorMapEntry for RevalidationSelectorAndHashes {
     }
 }
 
-/// Visitor to determine whether a selector requires cache revalidation.
-///
-/// Note that we just check simple selectors and eagerly return when the first
-/// need for revalidation is found, so we don't need to store state on the
-/// visitor.
-///
-/// Also, note that it's important to check the whole selector, due to cousins
-/// sharing arbitrarily deep in the DOM, not just the rightmost part of it
-/// (unfortunately, though).
-///
-/// With cousin sharing, we not only need to care about selectors in stuff like
-/// foo:first-child, but also about selectors like p:first-child foo, since the
-/// two parents may have shared style, and in that case we can test cousins
-/// whose matching depends on the selector up in the chain.
-///
-/// TODO(emilio): We can optimize when matching only siblings to only match the
-/// rightmost selector until a descendant combinator is found, I guess, and in
-/// general when we're sharing at depth `n`, to the `n + 1` sequences of
-/// descendant combinators.
-///
-/// I don't think that in presence of the bloom filter it's worth it, though.
-struct RevalidationVisitor;
+struct StylistSelectorVisitor<'a> {
+    needs_revalidation: bool,
+    passed_rightmost_selector: bool,
+    mapped_ids: &'a mut BloomFilter,
+    attribute_dependencies: &'a mut BloomFilter,
+    style_attribute_dependency: &'a mut bool,
+    state_dependencies: &'a mut ElementState,
+}
 
-impl SelectorVisitor for RevalidationVisitor {
-    type Impl = SelectorImpl;
+fn component_needs_revalidation(c: &Component<SelectorImpl>) -> bool {
```

Comment:
I am so glad the weird double-negative thing of the old setup has gone away.  I always had to detangle it very carefully.